x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/Mapper.java

```diff
@@ -202,6 +204,19 @@ private PhysicalPlan mapBinary(BinaryPlan bp) {
                     join.output()
                 );
             }
+            if (right instanceof FragmentExec fragment
+                && fragment.fragment() instanceof EsRelation relation
+                && relation.indexMode() == IndexMode.LOOKUP) {
+                return new LookupJoinExec(
+                    join.source(),
+                    left,
+                    right,
+                    config.matchFields(),
+                    config.leftFields(),
+                    config.rightFields(),
+                    join.output()
+                );
```

Comment:
I agree right now the abstraction is simplistic but the point is to decouple the command from the execution.

LookupJoin is syntactic sugar for special left join with special output - it gets folded into a regular join since I'd like to avoid serializing it (and dealing with versioning) and to have the compute engine pick up the pattern not the class.
That means any syntax that gets converted into a left join could be executed against a lookup join. Enrich for example could be attached onto this one but the dependency between the logical and physical happens through the pattern NOT the Lookup class.

Reply:
That's fair - but we aren't at the point yet where an Enrich, or anything else for that matter, is rewritten into a pattern that is translated into a LookupJoinExec. Currently, we only have the Lookup logical plan node, which doesn't need the additional round trip via surrogate + pattern matching.

I really prefer deferring such generalizations until they have actual use cases - because getting the generalization wrong early on leads to pain :/ and it makes the code more complex than it currently needs.

But that's supposed to be a nit anyway :)